### PR TITLE
Closing client when Proxyconnection is disconnected

### PR DESCRIPTION
**Pull request by jai1**
Currently, client.close() is not called when ProxyConnection becomes inactive which leads to leaking unused connections.

Since the ConnectionPool used in proxy uses the same event worker group as the Proxy service we can't close the event workgroup when we close the client, hence overriding the close() function in ProxyConnectionPool to just close the connections and not the event worker group.

### CODE REVIEW

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnectionPool.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package org.apache.pulsar.proxy.server;
```
Comment (massakam) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnectionPool.java]: There is no license header.

**Comment by massakam** [comment]
Don't we have to close `dnsResolver`?
https://github.com/apache/incubator-pulsar/blob/9348d34e830f558e3fc0ca096115bcc0d569ab0d/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionPool.java#L288-L292

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnectionPool.java (hunk 1 above)
Comment (jai1) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnectionPool.java]: Good catch - I will fix it

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 2]
```diff
@@ -140,6 +140,10 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
             directProxyHandler.outboundChannel.close();
         }
 
+        if (client != null) {
+            client.close();
+        }
```
Comment (merlimat) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: I don't understand why we want to close the entire client at this point. 

When this connection gets closed, we want its paired connection to be closed as well. This should already done at line 140 here above.

Comment (jai1) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: The client connection Pool is used by lookupProxyHandler and not the directProxyHandler.

Comment (merlimat) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: But the client is shared across all the ProxyConnection instances

Comment (jai1) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: We changed the logic to have one client per ProxyConnection 
https://github.com/apache/incubator-pulsar/pull/1707

Comment (merlimat) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: Uhm, that doesn't sound a good option to me

**Comment by jai1** [comment]
retest this please

**Comment by jai1** [comment]
retest this please

**Comment by jai1** [comment]
retest this please

**Comment by merlimat** [comment]
@jai1 there are many resources that are created per-client, I don't think it's a working approach to have 1 client instance per proxy connection.

**Comment by jai1** [comment]
I understand the resource aspect of it and agree that instead of having one client we can extract the connectionPool part out of the client and use one connection pool per proxy connection

But we still need one connection pool per proxy connection since the broker authenticates the client and the proxy if `forwardAuthorizationCredentials` is set to true and then the connection remains open throughout the lifetime of the proxy connection. 

So if Client A and Client B connect to Proxy we need one connection pool for A (since its A's credentials are authorized during handleConnect) and one connection pool for B.

If we are in agreement let me know and I will start the work of having one connection pool instead of one client per proxy connection.

**Comment by merlimat** [comment]
> But we still need one connection pool per proxy connection since the broker authenticates the client and the proxy if forwardAuthorizationCredentials is set to true and then the connection remains open throughout the lifetime of the proxy connection.

I think the term "connection pool" is actually misleading. Proxy uses a 1-1 connections matching between (client -> proxy) and (proxy -> broker). No pooling is required, but rather just opening new connection each time.

**Comment by jai1** [comment]
Ohk - now I understand the confusion 

The connection pool is used for lookups and getPartitionMetaData only

**Comment by sijie** [comment]
@jai1 @merlimat what is the conclusion here? are we going to include this for 2.1?

**Comment by jai1** [comment]
If this is blocking the release then we can defer it to the next one, else I will fix the tests and rebase the PR by Wednesday

**Comment by sijie** [comment]
@jai1 great! Wednesday is good.

**Comment by jai1** [comment]
@sijie - Can you review this PR so that we can merge it

**Comment by sijie** [comment]
@jai1 still looks good to me.

@merlimat can you review this? since you have some comments there.

**Comment by merlimat** [comment]
We can merge this now, though we should try to revisit the changes in #1707 for 2.2

**Pull request merged**